### PR TITLE
Adding weights to DC

### PR DESCRIPTION
**Pull request by shreyajain17**
To make it consistent with all other toolkits, DC should also have consistent batch_size in the entire epoch. This can be done by using weights. Weights are arrays of 0/1 of the size of targets. For example, the last batch in the epoch is smaller than the specified batch_size then weights are zeros for the rest of the batch. While calculating loss and accuracy, the data corresponding the zeros is not taken into consideration.

### CODE REVIEW

**Comment by jakesabathia2** [comment]
Seems like it crashes when evaluation?

**Comment by shreyajain17** [comment]
> Seems like it crashes when evaluation?

After your change evaluation change or before? Will look into it.

**Comment by jakesabathia2** [comment]
> > Seems like it crashes when evaluation?
> 
> After your evaluation change or before? Will look into it.

In Travis?
One unit test crash out.

**Comment by shreyajain17** [comment]
> > > Seems like it crashes when evaluation?
> > 
> > 
> > After your evaluation change or before? Will look into it.
> 
> In Travis?
> One unit test crash out.

Cool will look into it.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 1]
```diff
@@ -33,6 +33,15 @@ def __init__(self, net_params, batch_size, num_classes):
         self.batch_size = batch_size
 
         self.input = _tf.placeholder(_tf.float32, [None, 28, 28, 1])
```
Comment (jakesabathia2) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: If the batch size is fixed to a constant, why is there still has the shape None in the first dimension?
Like in OD the shape is fixed:
https://github.com/apple/turicreate/blob/dc07e46d4e5562f0989e8532feb735391061e170/src/python/turicreate/toolkits/object_detector/_tf_model_architecture.py#L47

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 2]
```diff
@@ -91,16 +100,12 @@ def __init__(self, net_params, batch_size, num_classes):
         self.predictions = softmax_out
 
         # Loss
-        self.cost = _tf.losses.softmax_cross_entropy(logits=out, onehot_labels=self.one_hot_labels, reduction=_tf.losses.Reduction.NONE)
+        self.cost = _tf.losses.softmax_cross_entropy(logits=out, onehot_labels=one_hot_labels, weights=reshaped_weights, reduction=_tf.losses.Reduction.NONE)
```
Comment (guihao-liang) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: if you can use pep8 format, that will be better for reviewing purposes.

**Comment by guihao-liang** [comment]
`While calculating loss and accuracy, the data corresponding to the zeros is not taken into consideration.`

Just curious why?

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 3]
```diff
@@ -192,41 +195,33 @@ data_iterator::batch simple_data_iterator::next_batch(size_t batch_size) {
   }
 
   if (target_index_ >= 0) {
-    DASSERT_EQ(real_batch_size, batch_targets.size());
+    batch_targets.resize(batch_size, 0.f);
```
Comment (guihao-liang) [inline src/toolkits/drawing_classifier/dc_data_iterator.cpp]: not sure why you don't want to use `real_batch_size`, which is quite critical.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 4]
```diff
@@ -746,7 +749,7 @@ gl_sframe drawing_classifier::perform_inference(data_iterator* data) const {
       pending_batches.pop();
 
       size_t num_images = batch.data_info.num_samples;
-      ASSERT_EQ(num_images * num_classes, batch.data_info.predictions.size());
+      ASSERT_EQ(batch_size * num_classes, batch.data_info.predictions.size());
```
Comment (guihao-liang) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: batch_size may not be real_batch_size, say you have 65 images, the batch size is 64, then you have 2 batches of 64 and 1.

Comment (shreyajain17) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Well the point of going this is that there will be two batch of 64. The batch size would never change. The weights are to weigh the loss. The weights for the first batch would all ones and for the second batch will be [1, 0, 0 ...]. The predictions come from the softmax would also be equal to the batch_size=64 but if you see the code after this, it will compute accuracy only till the num samples in the batch (which would be the real batch size)

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp (hunk 3 above)
Comment (shreyajain17) [reply src/toolkits/drawing_classifier/dc_data_iterator.cpp]: I am making the batch_targets to be of batch_size so assertion makes no sense to me.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 2 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Will do

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shreyajain17) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Makes sense.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 4 above)
Comment (guihao-liang) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: I see. What's the benefit of doing this? I'm curious, if the batch size is 64, and for the current batch, I only have 1 image, what you will do? Padding the rest 63 images?

Comment (guihao-liang) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Will this padding affect the batch norm?

Comment (shreyajain17) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: There is no batch norm in drawing classifier.

Comment (jakesabathia2) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Very good observe!
I also wondering why in our code we fix the batch size,
for instance, in AC we are doing the same thing and there are batch norm layers in the model.
And we might also change the model structure to something else for DC in the future (don't know this would happen tho haha), which might also include some cross-samples operation :).
But for now in the interest of consistent it makes sense to fix the batch size I guess haha.

Comment (guihao-liang) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: @shreyajain17 I mean for other toolkits with batch norms. I think this skewed batch norm will affect the final results.

**Comment by guihao-liang** [comment]
> es the friction/overhead of reviewing!

That's right! To alleviate the burden, I think we can prioritize #2312 and #2943.

I'm glad that _black_ helps @shreyajain17 on formatting!
